test/prober/sender.go

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020 The Knative Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prober
+
+import (
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/eventing/test/lib"
+)
+
+var senderName = "wathola-sender"
+
+func (p *prober) deploySender() {
+	p.log.Infof("Deploy sender pod: %v", senderName)
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      senderName,
+			Namespace: p.config.Namespace,
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name: configName,
+					VolumeSource: corev1.VolumeSource{
+						Secret: &corev1.SecretVolumeSource{
+							SecretName: configName,
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:  "sender",
+					Image: fmt.Sprintf("quay.io/cardil/wathola-sender:%v", Version),
```

Comment:
I didn't notice any code for the sender (and also receiver image). They should be added among the other test images: https://github.com/knative/eventing/tree/master/test/test_images , and should be built before the test

Reply:
I can do that, that's additional 1k of Golang code: https://github.com/cardil/wathola 

But, I don't see, why it must be embedded?! Having that as standalone tool gives you a lot of flexibility. For example @houshengbo ask here https://github.com/knative/eventing/issues/2271#issuecomment-573773102 if that code can be used in e2e tests and in operator e2e tests. Having a tool as separate entity ease that, as you can just vendor `test/prober` directory and run those tests in different context (also, in Openshift, etc.)

Reply:
> But, I don't see, why it must be embedded?! 

Because the code is important for understanding the whole test. Having 1k of Golang in some other repo doesn't help. You also mentioned the comment from @houshengbo . I don't see how having 1k lines of code in an external repo helps when you still have 7k lines of code in this repo. Should the tests be re-used in the Operator repo they would have to vendor from two different repos (this one and your private cardil/wathola). Also, the code should not be in a private repo!
All other tests which are using images have their code in test/test_images and we should continue doing it in this way.

Reply:
BTW, by having the code in your private repo makes you the only possible reviewer and no-one else can review/accept any changes to that code. The community looses control of that code.

Reply:
> Because the code is important for understanding the whole test. Having 1k of Golang in some other repo doesn't help.

I don't buy that. It's obvious what are the functions of _receiver_, _forwarder_, and _sender_, especially, after I've made a description of their functions in `test/upgrade/README.md`. Reading the code in `test/probe` and `test/upgrade` it's sufficient to understand what it is doing. And, still someone can read the code of `cardil/wathola` if he like to.

> I don't see how having 1k lines of code in an external repo helps

Here are just binary images being used. The code isn't built here. That's beneficial when trying to reuse that in other repos. One just vendor `test/probe` directory and can write its own test using that. 

If we embed that tool here, in other repos one would need also to implement image building/publishing from remote repo (this one). That is unnecessary complication and isn't done anywhere, I think.

> [..] they would have to vendor from two different repos (this one and your private cardil/wathola)

No, they don't! The code from `cardil/wathola` isn't used. Binary images are published to quay.io. In that sense it's just an external tool.

> Also, the code should not be in a private repo! [..] The community looses control of that code.

That not a valid argument. That code is still open source. Community can fork it, embed, do whatever they need. Or just use binaries like this PR does. Functionality of `cardil/wathola` is strictly defined by this task and will not change in foreseeable future. I don't see why it might need to change. 

But, when it does there are multiple ways of how community may process with that change. On top of that I'm completely happy to give away ownership of that repo to knative! I just think it should remain separate from `knative/eventing`.

Reply:
I think I'll open a PR to https://github.com/knative/test-infra repo to put code  from https://github.com/cardil/wathola to a `tools/` directory, with a caveat that deployment to image registry should be configured. In that way code like be under ownership of knative, but tools will be reachable without the need of building. 

Other good place to put it could be some repo under https://github.com/cloudevents organization.

I would like to retain the possibility of using pre-built images of wathola in different repos: here, in eventing-operator, in openshift's repos etc.

Reply:
That sounds good. And Dockerfiles for those images to https://github.com/knative/test-infra/tree/master/images ?

Reply:
Proposed in PR: https://github.com/knative/test-infra/pull/1631

Reply:
As of discussion on knative/test-infra#1631 I will embed that tool source code to this PR, making it even bigger. :dizzy_face:

Reply:
I propose I'll donate wathola testing tool to this repository in separate PR, not this one.

Adding it to this PR will make it even larger, but most importantly will make merging it harder. The new code will also need to pass review. That would lengthen the wait period for this PR, and to fix the issues it found.

What do you think @mgencur @vaikas @chizhg about that proposition?

Reply:
I too share similar concerns. Is there a reason why we can't use the more standard CloudEvents conformance tools:
https://github.com/cloudevents/conformance

@n3wscott

Reply:
I looked at https://github.com/cloudevents/conformance 

I don't see how that might be a replacement for wathola tool I built. I'm not only sending and receiving events. I also create a report to list all problems - events propagated more than once, missed events.

But, I can move missing functionality from wathola to https://github.com/cloudevents/conformance - I like that option. If that's okay with you...

Reply:
Yes, the conformance tool needs a ton of work still... 

the issue with quay is that we can't pull this code down and run the test without access to quay. The code needs to be built from source at the time of test I think for me to understand what is going to run on the cluster.

Reply:
Wathola tool is now incorporated in this PR. I no longer pull anything from quay.io